lib/backend.js

```diff
@@ -637,45 +641,120 @@ Backend.prototype._fetchSnapshot = function (collection, id, version, callback)
+        if (error) return callback(error);
 
-      for (var index = 0; index < ops.length; index++) {
-        var op = ops[index];
-        fetchedVersion = op.v + 1;
-
-        if (op.create) {
-          type = types.map[op.create.type];
-          if (!type) return callback({ code: 4008, message: 'Unknown type' });
-          data = type.create(op.create.data);
-        } else if (op.del) {
-          data = undefined;
-          type = null;
-        } else {
-          data = type.apply(data, op.op);
+        if (version > snapshot.v) {
+          return callback({ code: 4024, message: 'Requested version exceeds latest snapshot version' });
         }
-      }
 
-      type = type ? type.uri : null;
+        callback(null, snapshot);
+      });
+    });
+  });
+};
 
-      if (version > fetchedVersion) {
-        return callback({ code: 4024, message: 'Requested version exceeds latest snapshot version' });
-      }
+Backend.prototype.fetchSnapshotByTimestamp = function (agent, index, id, timestamp, callback) {
+  var start = Date.now();
+  var backend = this;
+  var projection = this.projections[index];
+  var collection = projection ? projection.target : index;
+  var request = {
+    agent: agent,
+    index: index,
+    collection: collection,
+    id: id,
+    timestamp: timestamp
+  };
 
-      var snapshot = new Snapshot(id, fetchedVersion, type, data, null);
+  this._fetchSnapshotByTimestamp(collection, id, timestamp, function (error, snapshot) {
+    if (error) return callback(error);
+    var snapshotProjection = backend._getSnapshotProjection(backend.db, projection);
+    var snapshots = [snapshot];
+    var snapshotType = backend.SNAPSHOT_TYPES.byTimestamp;
+    backend._sanitizeSnapshots(agent, snapshotProjection, collection, snapshots, snapshotType, function (error) {
+      if (error) return callback(error);
+      backend.emit('timing', 'fetchSnapshot', Date.now() - start, request);
       callback(null, snapshot);
     });
   });
 };
 
+Backend.prototype._fetchSnapshotByTimestamp = function (collection, id, timestamp, callback) {
+  var db = this.db;
+  var milestoneDb = this.milestoneDb;
+  var backend = this;
+
+  var milestoneSnapshot;
+  var from = 0;
+  var to = null;
+
+  milestoneDb.getMilestoneSnapshotAtOrBeforeTime(collection, id, timestamp, function (error, snapshot) {
+    if (error) return callback(error);
+    milestoneSnapshot = snapshot;
+    if (snapshot) from = snapshot.v;
+
+    milestoneDb.getMilestoneSnapshotAtOrAfterTime(collection, id, timestamp, function (error, snapshot) {
+      if (error) return callback(error);
+      if (snapshot) to = snapshot.v;
+
+      var options = {metadata: true};
+      db.getOps(collection, id, from, to, options, function (error, ops) {
+        if (error) return callback(error);
+        backend._buildSnapshotFromOps(id, milestoneSnapshot, ops, callback, function shouldBreak(nextOp) {
```

Comment:
As discussed, the `shouldBreak` was a bit confusing at first. Couple of options we came up with:

* Your idea of removing `shouldBreak`, instead pre-filtering the ops passed in.
* Passing in a `{maxOpTimestamp: ___}` options object.

I don't have a strong preference, as `_buildSnapshotFromOps` is a purely internal function.